hnn_qt5.py

```diff
@@ -218,71 +231,174 @@ def runsim (self):
-      import simdat
-      simdat.calcerr(simdat.ddat) # make sure error re-calculated (synchronously)
-      self.updatedrawerr() # send event to draw updated error (asynchronously)
+      self.runsim(is_opt=True) # run the simulation as usual and read its output
+
+      if self.last_step:
+        simdat.weighted_rmse(simdat.ddat,
+                            self.opt_params['opt_end'],
+                            self.opt_params['weights'],
+                            tstart=self.opt_params['opt_start'])
+        simdat.calcerr(simdat.ddat,
+                      self.opt_params['opt_end'],
+                      tstart=self.opt_params['opt_start'])
+        err = simdat.ddat['werrtot']
+#        print("regular RMSE = %f"% (simdat.ddat['errtot']))
+        print("weighted RMSE = %f, regular RMSE = %f"% (simdat.ddat['werrtot'],simdat.ddat['errtot']))
+      else:
+        simdat.calcerr(simdat.ddat,
+                      self.opt_params['opt_end'],
+                      tstart=self.opt_params['opt_start'])
+        simdat.weighted_rmse(simdat.ddat,
+                             self.opt_params['opt_end'],
+                             self.opt_params['weights'],
+                             tstart=self.opt_params['opt_start'])
+        err = simdat.ddat['werrtot']
+        print("weighted RMSE = %f, regular RMSE = %f"% (simdat.ddat['werrtot'],simdat.ddat['errtot']))
       self.updatewaitsimwin(os.linesep+'Simulation finished: error='+str(simdat.ddat['errtot'])+os.linesep) # print error
-      print(os.linesep+'Simulation finished: error='+str(simdat.ddat['errtot'])+os.linesep)#,'time=',time())
 
       with open(os.path.join(dconf['datdir'],paramf.split(os.path.sep)[-1].split('.param')[0],'optinf.txt'),'a') as fpopt:
         fpopt.write(str(simdat.ddat['errtot'])+os.linesep) # write error
 
       # backup the current param file
       outdir = os.path.join(dconf['datdir'],paramf.split(os.path.sep)[-1].split('.param')[0])
+      # last param file
       prmloc0 = os.path.join(outdir,paramf.split(os.path.sep)[-1])
-      prmloc1 = os.path.join(outdir,str(self.optiter)+'.param')
+
+      # save numbered by optiter
+      prmloc1 = os.path.join(outdir,'step_%d_iter_%d.param'%(self.cur_step,self.optiter))
       shutil.copyfile(prmloc0,prmloc1)
-      if simdat.ddat['errtot'] < self.minopterr:
-        self.minopterr = simdat.ddat['errtot']
-        shutil.copyfile(prmloc0,os.path.join(outdir,'best.param')) # convenience, save best here
+
+      if err < self.stepminopterr:
+        print("new best with RMSE %f"%err)
+        self.stepminopterr = err
+        # save best param file
+        shutil.copyfile(prmloc0,os.path.join(outdir,'step_%d_best.param'%self.cur_step)) # convenience, save best here
+        self.best_ddat = simdat.ddat.copy()
+
+      if self.optiter == 0 and not self.first_step:
+        # update plots for the first iteration only (best results from last round)
```

Comment:
Note that this doesn't happen on the first step because there are no optimization results to show.